### PR TITLE
Devnet optimizations

### PR DESCRIPTION
**Pull request by dribeiro-ShardLabs**
- Run Flask inside a proper container. This will improve how multiple requests are handled. Instead of being "piled up" increasing the time they take to process, the requests will now be received and treated sequentially.
- Add new optional CLI argument `--lite-mode`. When this argument is passed, devnet will not calculate transaction hashes for deploy transactions, and will instead use a numeric sequence, and it turns off block generation as well.

- `test_cli` and `test_cli_auth` were changed to `pytest` test files, in order to simplify unit test management;
- `test_postman` was changed to a single test function, as the features are meant to be tested in a single flow

### CODE REVIEW

File: starknet_devnet/server.py [hunk 1]
```diff
@@ -319,14 +319,15 @@ def main():
...
     dumper.dump_on = args.dump_on
-
-    app.run(host=args.host, port=args.port)
+    starknet_wrapper.lite_mode = args.lite_mode
+    try:
+        meinheld.listen((args.host, args.port))
+        meinheld.run(app)
+    finally:
+        if args.dump_on == DumpOn.EXIT:
+            dump_on_exit()
```
Comment (FabijanC) [inline starknet_devnet/server.py:330]: If `dump_on_exit` isn't used anywhere else, you can remove it and replace this call with `dumper.dump(dumper.dump_path)`.

**Comment by FabijanC** [comment]
Can you add dev changes to the PR description (like testing changes).

**Comment by dribeiro-ShardLabs** [comment]
> Can you add dev changes to the PR description (like testing changes)?

Done

File: test/test_dump.py [hunk 2]
```diff
@@ -70,63 +70,63 @@ def deploy_empty_contract():
     assert initial_balance == "0"
```
Comment (FabijanC) [inline test/test_dump.py:70]: Uncomment?

File: starknet_devnet/util.py [hunk 3]
```diff
@@ -123,6 +123,11 @@ def parse_args():
         help=f"Specify when to dump; can dump on: {DUMP_ON_OPTIONS_STRINGIFIED}",
         type=parse_dump_on
     )
+    parser.add_argument(
+        "--lite-mode",
```
Comment (FabijanC) [inline starknet_devnet/util.py:127]: How much work would it be to split this into multiple options? Like allowing for a more configurable approach with a separate `--blocks` and `--tx_hash` options. E.g. `--blocks lite`, `--blocks full`, `--tx_hash lite`, `--tx_hash real`?

File: test/test_cli_auth.py [hunk 4]
```diff
@@ -14,39 +15,59 @@
...
+]
+BALANCE_KEY = "142452623821144136554572927896792266630776240502820879601186867231282346767"
 
-assert_tx_status(deploy_info["tx_hash"], "ACCEPTED_ON_L2")
-assert_transaction(deploy_info["tx_hash"], "ACCEPTED_ON_L2")
+@pytest.fixture(autouse=True)
+def run_before_and_after_test():
+    """Run devnet before and kill it after the test run"""
+    # before test
+    devnet_proc = run_devnet_in_background(sleep_seconds=20)
```
Comment (FabijanC) [inline test/test_cli_auth.py:32]: Why so much sleep time?

File: starknet_devnet/starknet_wrapper.py [hunk 5]
```diff
@@ -179,13 +189,11 @@ async def invoke(self, transaction: InvokeFunction):
         invoke_transaction: InternalInvokeFunction = InternalInvokeFunction.from_external(transaction, state.general_config)
 
         try:
-            # This check might not be needed in future versions which will interact with the token contract
```
Comment (FabijanC) [inline starknet_devnet/starknet_wrapper.py:182]: Why did you delete this comment?

File: starknet_devnet/starknet_wrapper.py [hunk 6]
```diff
@@ -179,13 +189,11 @@ async def invoke(self, transaction: InvokeFunction):
...
             if invoke_transaction.max_fee: # handle only if non-zero
                 actual_fee = await self.calculate_actual_fee(transaction)
                 if actual_fee > invoke_transaction.max_fee:
                     message = f"Actual fee exceeded max fee.\n{actual_fee} > {invoke_transaction.max_fee}"
                     raise StarknetDevnetException(message=message)
-
```
Comment (FabijanC) [inline starknet_devnet/starknet_wrapper.py:188]: You can leave this empty line.

File: test/test_cli.py [hunk 7]
```diff
@@ -2,6 +2,8 @@
 The main testing script. Runs the devnet and calls its endpoints.
```
Comment (FabijanC) [inline test/test_cli.py:2]: As already noted in pyproject.toml comment, change the name of this and similar files.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: .pylintrc [hunk 8]
```diff
@@ -5,5 +5,12 @@ max-line-length=150
 min-public-methods=1
 
 [SIMILARITIES]
-min-similarity-lines=20
+min-similarity-lines=50
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply .pylintrc:8]: Pylint kept accusing similarities between test_cli, test_cli_auth and test_cli_lite

[the remark this answers is not in the document]
Comment (FabijanC) [reply .pylintrc:8]: Well it is kind of right, we should consider reusing commonalities, but probably not in this PR.

File: starknet_devnet/starknet_wrapper.py [hunk 9]
```diff
@@ -66,6 +66,10 @@ def __init__(self):
 
         self.__last_state_update = None
 
+        self.__tx_hash = 0
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply starknet_devnet/starknet_wrapper.py:74]: Why is a private method needed?

File: starknet_devnet/starknet_wrapper.py (hunk 5 above)
Comment (dribeiro-ShardLabs) [reply starknet_devnet/starknet_wrapper.py:182]: Rebasing issues

File: starknet_devnet/starknet_wrapper.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/starknet_wrapper.py:74]: I mean, I don't see a need for introducing a dedicated tx_hash counter when this same information is derivable from the number of thus far constructed transactions.

File: starknet_devnet/util.py (hunk 3 above)
Comment (dribeiro-ShardLabs) [reply starknet_devnet/util.py:127]: what about something like `--lite-mode-blocks` and `--lite-mode-tx-hash`, if a flag is passed the lite mode for it will be enabled

**Review by FabijanC — CHANGES_REQUESTED** [review]

**Review by FabijanC — APPROVED** [review]
